### ISSUE TITLE
ICE: `broken MIR: bad type assert (T = &T)` when declaring variable with let ref without assignment

### ISSUE THREAD

**Issue by hsandt** [rust-lang/rust#55308]
When writing a declaration with `let ref` and a type annotation but no assignment, compilation fails with a broken MIR error. Besides, a verbose build will hint at the whole block containing the faulty declaration, making it difficult to spot the faulty line.

I assume that this syntax is indeed invalid as there is nothing on the right side to refer to. However, I expected a localized and specific error message.

**Versions affected**

```
rustc 1.26.0 (a77568041 2018-05-07)
binary: rustc
commit-hash: a7756804103447ea4e68a71ccf071e7ad8f7a03e
commit-date: 2018-05-07
host: x86_64-unknown-linux-gnu
release: 1.26.0
LLVM version: 6.0

rustc 1.27.0-nightly (ff2ac35db 2018-05-12)
binary: rustc
commit-hash: ff2ac35db93a80b2de5daa4f280bf1503d62c164
commit-date: 2018-05-12
host: x86_64-unknown-linux-gnu
release: 1.27.0-nightly
LLVM version: 6.0
```

**MWE**

* Code

```
fn main() {
    let ref _i: i32;
}
```

* Compilation result

```
error: internal compiler error: broken MIR in NodeId(4) (UserAssertTy(Canonical { variables: Slice([]), value: i32 }, _1)): bad type assert (i32 = &i32): Sorts(ExpectedFound { expected: &i32, found: i32 })
 --> src/main.rs:1:11
  |
1 |   fn main() {
  |  ___________^
2 | |     let ref _i: i32;
3 | | }
  | |_^
```

* Expected

A dedicated error message should spot the faulty declaration line and explain that we cannot declare a variable with `let ref` without assignment.
Optionally, it may suggest that the user wanted to write `let _i: &i32` instead, but that's much more subjective (it may confuse the user by having him/her think that ref and the `&` in the type annotation are linked).

The declaration used to work in a previous version of Rust, so maybe it used to be an equivalent for `let _i: &i32`, but I only compiled the line without further usage so I never checked.

**Comment by pnkfelix** [comment]
Almost certainly not; I tended to add new checks, not remove preexisting ones ...

But I might know how to fix it...

**Comment by estebank** [comment]
[Can't repro on beta or nightly](https://play.rust-lang.org/?version=beta&mode=debug&edition=2015&gist=98bb37f3c7dd76ab024cf0eca022e73c).

**Comment by hsandt** [comment]
Indeed, my rust toolchains were old and the issue was fixed between 1.26 and 1.30... Sorry for the outdated issue.

As a side note, `let ref _i: i32;` is really an equivalent for `let _i: &i32;` now.
